README.md
# Coachella  

Coachella valley velocity model

## Installation

This package is intended to be installed as part of the UCVM framework,
version 19.4.0 or higher. 

## Library

The library ./lib/libcoachella.a may be statically linked into any
user application. Also, if your system supports dynamic linking,
you will also have a ./lib/libcoachella.so file that can be used
for dynamic linking. The header file defining the API is located
in ./include/coachella.h.

## Contact the authors

If you would like to contact the authors regarding this software,
please e-mail software@scec.org. Note this e-mail address should
be used for questions regarding the software itself (e.g. how
do I link the library properly?). Questions regarding the model's
science (e.g. on what paper is the COACHELLA valley velocity model 
based?) should be directed to the model's authors, located in the 
AUTHORS file.


ChangeLog
LSU Coachella Valley (2021):

	http://scec.usc.edu/scecpedia/LSUCV


tests/test.c
/**
 * @file test.c
 * @brief Bootstraps the test framework for the COACHELLA library.
 * @author - SCEC
 * @version 1.0
 *
 * Tests the ALBACORE library by loading it and executing the code as
 * UCVM would do it.
 *
 */

#include <stdlib.h>
#include <stdio.h>
#include <assert.h>
#include "coachella.h"

/**
 * Initializes and runs the test program. Tests link against the
 * static version of the library to prevent any dynamic loading
 * issues.
 *
 * @param argc The number of arguments.
 * @param argv The argument strings.
 * @return A zero value indicating success.
 */
int main(int argc, const char* argv[]) {

	// Declare the structures.
	coachella_point_t pt;
	coachella_properties_t ret;

	// Initialize the model.
	assert(coachella_init("../", "coachella") == 0);

	printf("Loaded the model successfully.\n");

	// Query a point.
	pt.longitude = -118;
	pt.latitude = 34;
	pt.depth = 0;

	coachella_query(&pt, &ret, 1);

	assert(ret.vs > 0);
	assert(ret.vp > 0);
	assert(ret.rho > 0);

	printf("Query was successful.\n");

	// Close the model.
	assert(coachella_finalize() == 0);

	printf("Model closed successfully.\n");

	printf("\nALL COACHELLA TESTS PASSED");

	return 0;
}


src/coachella.h
/**
 * @file coachella.h
 * @brief Main header file for LSU Coachella Valley library.
 * @author - SCEC 
 * @version 1.0
 *
 * Delivers LSU Coachella Valley Velocity Model
 *
 */

// Includes
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <math.h>

// Constants
#ifndef M_PI
	/** Defines pi */
	#define M_PI 3.14159265358979323846
#endif

/** Defines a return value of success */
#define SUCCESS 0
/** Defines a return value of failure */
#define FAIL 1
/** Defines a return value of NA from model */
#define NA -1 

// Structures
/** Defines a point (latitude, longitude, and depth) in WGS84 format */
typedef struct coachella_point_t {
	/** Longitude member of the point */
	double longitude;
	/** Latitude member of the point */
	double latitude;
	/** Depth member of the point */
	double depth;
} coachella_point_t;

/** Defines the material properties this model will retrieve. */
typedef struct coachella_properties_t {
	/** P-wave velocity in meters per second */
	double vp;
	/** S-wave velocity in meters per second */
	double vs;
	/** Density in g/m^3 */
	double rho;
} coachella_properties_t;

/** The COACHELLA configuration structure. */
typedef struct coachella_configuration_t {
	/** The zone of UTM projection */
	int utm_zone;
	/** The model directory */
	char model_dir[128];
	/** Number of x points */
	int nx;
	/** Number of y points */
	int ny;
	/** Number of z points */
	int nz;
	/** Depth in meters */
	double depth;
	/** Top left corner easting */
	double top_left_corner_lon;
	/** Top left corner northing */
	double top_left_corner_lat;
	/** Top right corner easting */
	double top_right_corner_lon;
	/** Top right corner northing */
	double top_right_corner_lat;
	/** Bottom left corner easting */
	double bottom_left_corner_lon;
	/** Bottom left corner northing */
	double bottom_left_corner_lat;
	/** Bottom right corner easting */
	double bottom_right_corner_lat;
	/** Bottom right corner northing */
	double bottom_right_corner_lon;
	/** Z interval for the data */
	double depth_interval;
        /** Bilinear or Trilinear Interpolation on or off (1 or 0) */
        int interpolation;

} coachella_configuration_t;

/** The model structure which points to available portions of the model. */
typedef struct coachella_model_t {
	/** A pointer to the Vp data either in memory or disk. Null if does not exist. */
	void *vp;
	/** Vp status: 0 = not found, 1 = found and not in memory, 2 = found and in memory */
	int vp_status;
} coachella_model_t;

// Constants
/** The version of the model. */
const char *coachella_version_string = "COACHELLA";

// Variables
/** Set to 1 when the model is ready for query. */
int coachella_is_initialized = 0;

/** Location of the binary data files. */
char coachella_data_directory[128];

/** Configuration parameters. */
coachella_configuration_t *coachella_configuration;
/** Holds pointers to the velocity model data OR indicates it can be read from file. */
coachella_model_t *coachella_velocity_model;

/** The height of this model's region, in meters. */
double coachella_total_height_m = 0;
/** The width of this model's region, in meters. */
double coachella_total_width_m = 0;

// UCVM API Required Functions

#ifdef DYNAMIC_LIBRARY

/** Initializes the model */
int model_init(const char *dir, const char *label);
/** Cleans up the model (frees memory, etc.) */
int model_finalize();
/** Returns version information */
int model_version(char *ver, int len);
/** Queries the model */
int model_query(coachella_point_t *points, coachella_properties_t *data, int numpts);

#endif

// COACHELLA Related Functions

/** Initializes the model */
int coachella_init(const char *dir, const char *label);
/** Cleans up the model (frees memory, etc.) */
int coachella_finalize();
/** Returns version information */
int coachella_version(char *ver, int len);
/** Queries the model */
int coachella_query(coachella_point_t *points, coachella_properties_t *data, int numpts);

// Non-UCVM Helper Functions
/** Reads the configuration file. */
int coachella_read_configuration(char *file, coachella_configuration_t *config);
void print_error(char *err);
/** Retrieves the value at a specified grid point in the model. */
void coachella_read_properties(int x, int y, int z, coachella_properties_t *data);
/** Attempts to malloc the model size in memory and read it in. */
int coachella_try_reading_model(coachella_model_t *model);
/** Calculates density from Vs. */
double coachella_calculate_density(double vp);
/** Calculates Vs from Vp. */
double coachella_calculate_vs(double vp);

// Interpolation Functions
/** Linearly interpolates two coachella_properties_t structures */
void coachella_linear_interpolation(double percent, coachella_properties_t *x0, coachella_properties_t *x1, coachella_properties_t *ret_properties);
/** Bilinearly interpolates the properties. */
void coachella_bilinear_interpolation(double x_percent, double y_percent, coachella_properties_t *four_points, coachella_properties_t *ret_properties);
/** Trilinearly interpolates the properties. */
void coachella_trilinear_interpolation(double x_percent, double y_percent, double z_percent, coachella_properties_t *eight_points,
							 coachella_properties_t *ret_properties);


data/make_data_file.py
#!/usr/bin/env python


##
#  Builds the data files in the expected format from CV_3D_Model.txt
#
# from >> lon lat depth(m) vp(km/s)
#
# to >>  /** P-wave velocity in km/s per second */
#        double vp;
# depth is in increment of 500m,
#
 

import getopt
import sys
import subprocess
import struct
import array

## at hypocenter  LSU/CV_3D_Model.txt

model = "LSU"

#174629
dimension_x = 101
dimension_y = 91
dimension_z = 19

lon_origin = -116.7
lat_origin = 33.3

lon_upper = -115.7
lat_upper = 34.2


delta_lon = (lon_upper - lon_origin )/(dimension_x-1)
delta_lat = (lat_upper - lat_origin)/(dimension_y-1)

for ycoord in xrange(dimension_y):
   yloc = lat_origin + (ycoord * delta_lat)
#  print "yloc - ",yloc
for xcoord in xrange(dimension_x):
   xloc = lon_origin + (xcoord * delta_lon)
#   print  "xloc - ",xloc

def usage():
    print("\n./make_data_files.py -u [uid]\n\n")
    print("-u - username to use to do the dataset retrieval.\n")
    sys.exit(0)

def main():

    # Set our variable defaults.
    username = ""
    path = "/var/www/html/research/ucvmc/" + model 

    try:
        opts, args = getopt.getopt(sys.argv[1:], "u:", ["user="])
    except getopt.GetoptError as err:
        print(str(err))
        usage()
        sys.exit(1)

    for o, a in opts:
        if o in ("-u", "--user"):
            username = str(a) + "@"

##
##    subprocess.check_call(["scp", username +
##                           "hypocenter.usc.edu:" + path + "/CV_3D_Model.txt",
##                           "."])
##

    # Now we need to go through the data files and put them in the correct
    # format for LSUCV. More specifically, we need a Vp.dat

    print("\nWriting out LSUCV data files\n")

    f = open("./CV_3D_Model.txt")

    f_vp = open("./cv/vp.dat", "wb")

#    print("size of model, ",dimension_x * dimension_y * dimension_z)

    vp_arr = array.array('f', (-1.0,) * (dimension_x * dimension_y * dimension_z))

    nan_cnt = 0
    total_cnt =0;
    for line in f:
        arr = line.split()
        lon_v = float(arr[0])
        lat_v = float(arr[1])
        depth_v = float(arr[2])/500
        vp = -1.0;
        tmp = arr[3]
        total_cnt = total_cnt + 1
        if( tmp != "NaN" ):
          vp = float(arr[3])
          vp = vp * 1000.0;
        else:
          nan_cnt = nan_cnt + 1

        y_pos = int(round((lat_v - lat_origin) / delta_lat))
        x_pos = int(round((lon_v - lon_origin) / delta_lon))
        z_pos = int(depth_v)

        loc =z_pos * (dimension_y * dimension_x) + (y_pos * dimension_x) + x_pos

        vp_arr[loc] = vp
#        if( tmp != "NaN" ):
#          print x_pos," ",y_pos," ",z_pos," >> ",lon_v," ",lat_v," ",depth_v," ",vp 
#        else:
#          print "NAN", x_pos," ",y_pos," ",z_pos," >> ",lon_v," ",lat_v," ",depth_v," ",vp 

    vp_arr.tofile(f_vp)

    f.close()
    f_vp.close()

    print("Done! with NaN", nan_cnt, "total", total_cnt)

if __name__ == "__main__":
    main()



data/query_data_file.py
#!/usr/bin/env python

##
#  Retrieve the latlons form the original data txt file and
#  and retrieve the material properties from binary vp.dat file 
#

import getopt
import sys
import subprocess
import struct
import numpy as np

dimension_x = 101 
dimension_y = 91
dimension_z = 19

lon_origin = -116.7
lat_origin = 33.3

lon_upper = -115.7
lat_upper = 34.2

delta_lon = (lon_upper - lon_origin )/(dimension_x-1)
delta_lat = (lat_upper - lat_origin)/(dimension_y-1)

def usage():
    print("\n./query_data_files.py\n\n")
    sys.exit(0)

def main():

    count =0

    f_lonlat = open("./CV_3D_Model.txt")

    f_vp = open("./cv/vp.dat")

    vp_arr = np.fromfile(f_vp, dtype=np.float32)

    f_vp.close()

    lon_start = lon_origin
    lat_start = lat_origin
    depth_start = 0.0;

    count =0
    for line in f_lonlat:
        arr = line.split()
        lon_v = float(arr[0])
        lat_v = float(arr[1])
        depth_v = float(arr[2])/500
        fvp = arr[3]

        y_pos = int(round((lat_v - lat_origin) / delta_lat))
        x_pos = int(round((lon_v - lon_origin) / delta_lon))
        z_pos = int(depth_v)

        offset=z_pos * (dimension_y * dimension_x) + (y_pos * dimension_x) + x_pos
        vp=vp_arr[offset];

        if((fvp != "NaN") & (count < 10)) :
          count = count + 1
          print "xyz:", x_pos," ",y_pos," ",z_pos," >> ", lon_v, " ",lat_v, " ", float(depth_v) , "-->vp", vp, "->fvp", fvp
          print "offset ",offset

        if((float(vp) != -1) & ( (float(vp) > 8000) | (float(vp)<1500))) :
          print lon_v, " ",lat_v, " ", float(depth_v) , "-->vp", vp

    f_lonlat.close()
    print("\nDone!")

if __name__ == "__main__":
    main()




src/coachella.c
/**
 * @file coachella.c
 * @brief Main file for COACHELLA-LSU library.
 * @author - SCEC 
 * @version 1.0
 *
 * @section DESCRIPTION
 *
 * Delivers LSU Coachella Valley Velocity Model
 *
 */

#include "coachella.h"

/**
 * Initializes the COACHELLA plugin model within the UCVM framework. In order to initialize
 * the model, we must provide the UCVM install path and optionally a place in memory
 * where the model already exists.
 *
 * @param dir The directory in which UCVM has been installed.
 * @param label A unique identifier for the velocity model.
 * @return Success or failure, if initialization was successful.
 */
int coachella_init(const char *dir, const char *label) {
	int tempVal = 0;
	char configbuf[512];

	// Initialize variables.
	coachella_configuration = calloc(1, sizeof(coachella_configuration_t));
	coachella_velocity_model = calloc(1, sizeof(coachella_model_t));

	// Configuration file location.
	sprintf(configbuf, "%s/model/%s/data/config", dir, label);

	// Read the configuration file.
	if (coachella_read_configuration(configbuf, coachella_configuration) != SUCCESS)
		tempVal = FAIL;

	// Set up the data directory.
	sprintf(coachella_data_directory, "%s/model/%s/data/%s/", dir, label, coachella_configuration->model_dir);

	// Can we allocate the model, or parts of it, to memory. If so, we do.
	tempVal = coachella_try_reading_model(coachella_velocity_model);

	if (tempVal == SUCCESS) {
		fprintf(stderr, "WARNING: Could not load model into memory. Reading the model from the\n");
		fprintf(stderr, "hard disk may result in slow performance.");
	} else if (tempVal == FAIL) {
		print_error("No model file was found to read from.");
		return FAIL;
	}

	// In order to simplify our calculations in the query, we want to rotate the box so that the bottom-left
	// corner is at (0m,0m). Our box's height is total_height_m and total_width_m. We then rotate the
	// point so that is is somewhere between (0,0) and (total_width_m, total_height_m). How far along
	// the X and Y axis determines which grid points we use for the interpolation routine.

	coachella_total_height_m = sqrt(pow(coachella_configuration->top_left_corner_lat - coachella_configuration->bottom_left_corner_lat, 2.0f) +
						  pow(coachella_configuration->top_left_corner_lon - coachella_configuration->bottom_left_corner_lon, 2.0f));
	coachella_total_width_m  = sqrt(pow(coachella_configuration->top_right_corner_lat - coachella_configuration->top_left_corner_lat, 2.0f) +
						  pow(coachella_configuration->top_right_corner_lon - coachella_configuration->top_left_corner_lon, 2.0f));

	// Let everyone know that we are initialized and ready for business.
	coachella_is_initialized = 1;

	return SUCCESS;
}

/**
 * Queries COACHELLA at the given points and returns the data that it finds.
 *
 * @param points The points at which the queries will be made.
 * @param data The data that will be returned (Vp, Vs, density, Qs, and/or Qp).
 * @param numpoints The total number of points to query.
 * @return SUCCESS or FAIL.
 */
int coachella_query(coachella_point_t *points, coachella_properties_t *data, int numpoints) {
	int i = 0;
	int load_x_coord = 0, load_y_coord = 0, load_z_coord = 0;
	double x_percent = 0, y_percent = 0, z_percent = 0;
	coachella_properties_t surrounding_points[8];
        double lon_e, lat_n;

	int zone = 11;
	int longlat2utm = 0;

        double delta_lon = (coachella_configuration->top_right_corner_lon - coachella_configuration->bottom_left_corner_lon)/(coachella_configuration->nx - 1);
        double delta_lat = (coachella_configuration->top_right_corner_lat - coachella_configuration->bottom_left_corner_lat)/(coachella_configuration->ny - 1);

	for (i = 0; i < numpoints; i++) {
		lon_e = points[i].longitude; 
		lat_n = points[i].latitude; 

		// Which point base point does that correspond to?
		load_y_coord = (int)(round((lat_n - coachella_configuration->bottom_left_corner_lat) / delta_lat));
		load_x_coord = (int)(round((lon_e - coachella_configuration->bottom_left_corner_lon) / delta_lon));
		load_z_coord = (int)((points[i].depth)/1000);

		// Are we outside the model's X and Y and Z boundaries?
		if (points[i].depth > coachella_configuration->depth || load_x_coord > coachella_configuration->nx -1  || load_y_coord > coachella_configuration->ny -1 || load_x_coord < 0 || load_y_coord < 0 || load_z_coord < 0) {
			data[i].vp = -1;
			data[i].vs = -1;
			data[i].rho = -1;
			continue;
		}

		// Get the X, Y, and Z percentages for the bilinear or trilinear interpolation below.
		x_percent =fmod((lon_e - coachella_configuration->bottom_left_corner_lon), delta_lon) /delta_lon;
		y_percent = fmod((lat_n - coachella_configuration->bottom_left_corner_lat), delta_lat)/
delta_lat;
		z_percent = fmod(points[i].depth, coachella_configuration->depth_interval) / coachella_configuration->depth_interval;

		if (load_z_coord == 0 && z_percent == 0) {
			// We're below the model boundaries. Bilinearly interpolate the bottom plane and use that value.
			load_z_coord = 0;
                   if(coachella_configuration->interpolation) {

			// Get the four properties.
			coachella_read_properties(load_x_coord,     load_y_coord,     load_z_coord,     &(surrounding_points[0]));	// Orgin.
			coachella_read_properties(load_x_coord + 1, load_y_coord,     load_z_coord,     &(surrounding_points[1]));	// Orgin + 1x
			coachella_read_properties(load_x_coord,     load_y_coord + 1, load_z_coord,     &(surrounding_points[2]));	// Orgin + 1y
			coachella_read_properties(load_x_coord + 1, load_y_coord + 1, load_z_coord,     &(surrounding_points[3]));	// Orgin + x + y, forms top plane.

			coachella_bilinear_interpolation(x_percent, y_percent, surrounding_points, &(data[i]));
                  } else {
			coachella_read_properties(load_x_coord,     load_y_coord,     load_z_coord,     &(data[i]));	// Orgin.
                  }

		} else {
		  if( coachella_configuration->interpolation) {
			// Read all the surrounding point properties.
			coachella_read_properties(load_x_coord,     load_y_coord,     load_z_coord,     &(surrounding_points[0]));	// Orgin.
			coachella_read_properties(load_x_coord + 1, load_y_coord,     load_z_coord,     &(surrounding_points[1]));	// Orgin + 1x
			coachella_read_properties(load_x_coord,     load_y_coord + 1, load_z_coord,     &(surrounding_points[2]));	// Orgin + 1y
			coachella_read_properties(load_x_coord + 1, load_y_coord + 1, load_z_coord,     &(surrounding_points[3]));	// Orgin + x + y, forms top plane.
			coachella_read_properties(load_x_coord,     load_y_coord,     load_z_coord - 1, &(surrounding_points[4]));	// Bottom plane origin
			coachella_read_properties(load_x_coord + 1, load_y_coord,     load_z_coord - 1, &(surrounding_points[5]));	// +1x
			coachella_read_properties(load_x_coord,     load_y_coord + 1, load_z_coord - 1, &(surrounding_points[6]));	// +1y
			coachella_read_properties(load_x_coord + 1, load_y_coord + 1, load_z_coord - 1, &(surrounding_points[7]));	// +x +y, forms bottom plane.

			coachella_trilinear_interpolation(x_percent, y_percent, z_percent, surrounding_points, &(data[i]));
                    } else {
                        // no interpolation, data as it is
			coachella_read_properties(load_x_coord,     load_y_coord,     load_z_coord,     &(data[i]));	// Orgin.
                    }
		}

		data[i].rho = coachella_calculate_density(data[i].vp);
		data[i].vs = coachella_calculate_vs(data[i].vp);
	}

	return SUCCESS;
}

/**
 * Retrieves the material properties (whatever is available) for the given data point, expressed
 * in x, y, and z co-ordinates.
 *
 * @param x The x coordinate of the data point.
 * @param y The y coordinate of the data point.
 * @param z The z coordinate of the data point.
 * @param data The properties struct to which the material properties will be written.
 */
void coachella_read_properties(int x, int y, int z, coachella_properties_t *data) {

	// Set everything to -1 to indicate not found.
	data->vp = -1;
	data->vs = -1;
	data->rho = -1;

	float *ptr = NULL;
	FILE *fp = NULL;

	int location = z * (coachella_configuration->nx * coachella_configuration->ny) + (y * coachella_configuration->nx) + x;

	// Check our loaded components of the model.
	if (coachella_velocity_model->vp_status == 2) {
		// Read from memory.
		ptr = (float *)coachella_velocity_model->vp;
		data->vp = ptr[location];
	} else if (coachella_velocity_model->vp_status == 1) {
		// Read from file.
		fseek(fp, location * sizeof(float), SEEK_SET);
		fread(&(data->vp), sizeof(float), 1, fp);
	}
}

/**
 * Trilinearly interpolates given a x percentage, y percentage, z percentage and a cube of
 * data properties in top origin format (top plane first, bottom plane second).
 *
 * @param x_percent X percentage
 * @param y_percent Y percentage
 * @param z_percent Z percentage
 * @param eight_points Eight surrounding data properties
 * @param ret_properties Returned data properties
 */
void coachella_trilinear_interpolation(double x_percent, double y_percent, double z_percent,
							 coachella_properties_t *eight_points, coachella_properties_t *ret_properties) {
	coachella_properties_t *temp_array = calloc(2, sizeof(coachella_properties_t));
	coachella_properties_t *four_points = eight_points;

	coachella_bilinear_interpolation(x_percent, y_percent, four_points, &temp_array[0]);

	// Now advance the pointer four "coachella_properties_t" spaces.
	four_points += 4;

	// Another interpolation.
	coachella_bilinear_interpolation(x_percent, y_percent, four_points, &temp_array[1]);

	// Now linearly interpolate between the two.
	coachella_linear_interpolation(z_percent, &temp_array[0], &temp_array[1], ret_properties);

	free(temp_array);
}

/**
 * Bilinearly interpolates given a x percentage, y percentage, and a plane of data properties in
 * origin, bottom-right, top-left, top-right format.
 *
 * @param x_percent X percentage.
 * @param y_percent Y percentage.
 * @param four_points Data property plane.
 * @param ret_properties Returned data properties.
 */
void coachella_bilinear_interpolation(double x_percent, double y_percent, coachella_properties_t *four_points, coachella_properties_t *ret_properties) {

	coachella_properties_t *temp_array = calloc(2, sizeof(coachella_properties_t));

	coachella_linear_interpolation(x_percent, &four_points[0], &four_points[1], &temp_array[0]);
	coachella_linear_interpolation(x_percent, &four_points[2], &four_points[3], &temp_array[1]);
	coachella_linear_interpolation(y_percent, &temp_array[0], &temp_array[1], ret_properties);

	free(temp_array);
}

/**
 * Linearly interpolates given a percentage from x0 to x1, a data point at x0, and a data point at x1.
 *
 * @param percent Percent of the way from x0 to x1 (from 0 to 1 interval).
 * @param x0 Data point at x0.
 * @param x1 Data point at x1.
 * @param ret_properties Resulting data properties.
 */
void coachella_linear_interpolation(double percent, coachella_properties_t *x0, coachella_properties_t *x1, coachella_properties_t *ret_properties) {

	ret_properties->vp  = (1 - percent) * x0->vp  + percent * x1->vp;
	ret_properties->vs  = (1 - percent) * x0->vs  + percent * x1->vs;
	ret_properties->rho = (1 - percent) * x0->rho + percent * x1->rho;
}

/**
 * Called when the model is being discarded. Free all variables.
 *
 * @return SUCCESS
 */
int coachella_finalize() {
	if (coachella_velocity_model->vp) free(coachella_velocity_model->vp);

	free(coachella_configuration);

	return SUCCESS;
}

/**
 * Returns the version information.
 *
 * @param ver Version string to return.
 * @param len Maximum length of buffer.
 * @return Zero
 */
int coachella_version(char *ver, int len)
{
  int verlen;
  verlen = strlen(coachella_version_string);
  if (verlen > len - 1) {
    verlen = len - 1;
  }
  memset(ver, 0, len);
  strncpy(ver, coachella_version_string, verlen);
  return 0;
}

/**
 * Reads the configuration file describing the various properties of CVM-S5 and populates
 * the configuration struct. This assumes configuration has been "calloc'ed" and validates
 * that each value is not zero at the end.
 *
 * @param file The configuration file location on disk to read.
 * @param config The configuration struct to which the data should be written.
 * @return Success or failure, depending on if file was read successfully.
 */
int coachella_read_configuration(char *file, coachella_configuration_t *config) {
	FILE *fp = fopen(file, "r");
	char key[40];
	char value[80];
	char line_holder[128];

	// If our file pointer is null, an error has occurred. Return fail.
	if (fp == NULL) {
		print_error("Could not open the configuration file.");
		return FAIL;
	}

	// Read the lines in the configuration file.
	while (fgets(line_holder, sizeof(line_holder), fp) != NULL) {
		if (line_holder[0] != '#' && line_holder[0] != ' ' && line_holder[0] != '\n') {
			sscanf(line_holder, "%s = %s", key, value);

			// Which variable are we editing?
			if (strcmp(key, "utm_zone") == 0)
  				config->utm_zone = atoi(value);
			if (strcmp(key, "model_dir") == 0)
				sprintf(config->model_dir, "%s", value);
			if (strcmp(key, "nx") == 0)
  				config->nx = atoi(value);
			if (strcmp(key, "ny") == 0)
  			 	config->ny = atoi(value);
			if (strcmp(key, "nz") == 0)
  			 	config->nz = atoi(value);
			if (strcmp(key, "depth") == 0)
  			 	config->depth = atof(value);
			if (strcmp(key, "top_left_corner_lon") == 0)
				config->top_left_corner_lon = atof(value);
			if (strcmp(key, "top_left_corner_lat") == 0)
		 		config->top_left_corner_lat = atof(value);
			if (strcmp(key, "top_right_corner_lon") == 0)
				config->top_right_corner_lon = atof(value);
			if (strcmp(key, "top_right_corner_lat") == 0)
				config->top_right_corner_lat = atof(value);
			if (strcmp(key, "bottom_left_corner_lon") == 0)
				config->bottom_left_corner_lon = atof(value);
			if (strcmp(key, "bottom_left_corner_lat") == 0)
				config->bottom_left_corner_lat = atof(value);
			if (strcmp(key, "bottom_right_corner_lon") == 0)
				config->bottom_right_corner_lon = atof(value);
			if (strcmp(key, "bottom_right_corner_lat") == 0)
				config->bottom_right_corner_lat = atof(value);
			if (strcmp(key, "depth_interval") == 0)
				config->depth_interval = atof(value);
			if (strcmp(key, "interpolation") == 0) {
                                if (strcmp(value, "on") == 0) {
                                     config->interpolation = 1;
                                     } else {
                                          config->interpolation = 0;
                                }
                        };

		}
	}

	// Have we set up all configuration parameters?
	if (config->utm_zone == 0 || config->nx == 0 || config->ny == 0 || config->nz == 0 || config->model_dir[0] == '\0' ||
		config->top_left_corner_lon == 0 || config->top_left_corner_lat == 0 || config->top_right_corner_lon == 0 ||
		config->top_right_corner_lat == 0 || config->bottom_left_corner_lon == 0 || config->bottom_left_corner_lat == 0 ||
		config->bottom_right_corner_lon == 0 || config->bottom_right_corner_lat == 0 || config->depth == 0 ||
		config->depth_interval == 0) {
		print_error("One configuration parameter not specified. Please check your configuration file.");
		return FAIL;
	}

	fclose(fp);

	return SUCCESS;
}

/**
 * Calculates the density based off of Vp. Base on Brocher's formulae
 *
 * @param vp 
 * @return Density, in g/m^3.
 * [eqn. 6] r (g/cm3) = 1.6612Vp – 0.4721Vp2 + 0.0671Vp3 – 0.0043Vp4 + 0.000106Vp5.
 * Equation 6 is the “Nafe-Drake curve” (Ludwig et al., 1970).
 * start with vp in km 
 */
double coachella_calculate_density(double vp) {
     double retVal ;

     vp = vp * 0.001;
     retVal = vp * (1.6612 - vp * (0.4721 - vp * (0.0671 - vp * (0.0043 - vp * 0.000106))));
     if (retVal < 1.0) {
       retVal = 1.0;
     }
     retVal = retVal * 1000.0;
     return retVal;
}

/**
 * Calculates the vs based off of Vp. Base on Brocher's formulae
 *
 * https://pubs.usgs.gov/of/2005/1317/of2005-1317.pdf
 *
 * @param vp
 * @return Vs, in km.
 * Vs derived from Vp, Brocher (2005) eqn 1.
 * [eqn. 1] Vs (km/s) = 0.7858 – 1.2344Vp + 0.7949Vp2 – 0.1238Vp3 + 0.0064Vp4.
 * Equation 1 is valid for 1.5 < Vp < 8 km/s.
 */
double coachella_calculate_vs(double vp) {
     double retVal ;

     vp = vp * 0.001;
     retVal = 0.7858 - (vp *(1.2344 + vp *( 0.7949 - vp *(0.1238 + vp * 0.0064))));
     retVal = retVal * 1000.0;
     return retVal;
}


/**
 * Prints the error string provided.
 *
 * @param err The error string to print out to stderr.
 */
void print_error(char *err) {
	fprintf(stderr, "An error has occurred while executing COACHELLA. The error was:\n\n");
	fprintf(stderr, "%s", err);
	fprintf(stderr, "\n\nPlease contact software@scec.org and describe both the error and a bit\n");
	fprintf(stderr, "about the computer you are running COACHELLA on (Linux, Mac, etc.).\n");
}

/**
 * Tries to read the model into memory.
 *
 * @param model The model parameter struct which will hold the pointers to the data either on disk or in memory.
 * @return 2 if all files are read to memory, SUCCESS if file is found but at least 1
 * is not in memory, FAIL if no file found.
 */
int coachella_try_reading_model(coachella_model_t *model) {
	double base_malloc = coachella_configuration->nx * coachella_configuration->ny * coachella_configuration->nz * sizeof(float);
	int file_count = 0;
	int all_read_to_memory = 1;
	char current_file[128];
	FILE *fp;

	// Let's see what data we actually have.
	sprintf(current_file, "%s/vp.dat", coachella_data_directory);
	if (access(current_file, R_OK) == 0) {
		model->vp = malloc(base_malloc);
		if (model->vp != NULL) {
			// Read the model in.
			fp = fopen(current_file, "rb");
			fread(model->vp, 1, base_malloc, fp);
			fclose(fp);
			model->vp_status = 2;
		} else {
			all_read_to_memory = 0;
			model->vp = fopen(current_file, "rb");
			model->vp_status = 1;
		}
		file_count++;
	}

	if (file_count == 0)
		return FAIL;
	else if (file_count > 0 && all_read_to_memory == 0)
		return SUCCESS;
	else
		return 2;
}

// The following functions are for dynamic library mode. If we are compiling
// a static library, these functions must be disabled to avoid conflicts.
#ifdef DYNAMIC_LIBRARY

/**
 * Init function loaded and called by the UCVM library. Calls coachella_init.
 *
 * @param dir The directory in which UCVM is installed.
 * @return Success or failure.
 */
int model_init(const char *dir, const char *label) {
	return coachella_init(dir, label);
}

/**
 * Query function loaded and called by the UCVM library. Calls coachella_query.
 *
 * @param points The basic_point_t array containing the points.
 * @param data The basic_properties_t array containing the material properties returned.
 * @param numpoints The number of points in the array.
 * @return Success or fail.
 */
int model_query(coachella_point_t *points, coachella_properties_t *data, int numpoints) {
	return coachella_query(points, data, numpoints);
}

/**
 * Finalize function loaded and called by the UCVM library. Calls coachella_finalize.
 *
 * @return Success
 */
int model_finalize() {
	return coachella_finalize();
}

/**
 * Version function loaded and called by the UCVM library. Calls coachella_version.
 *
 * @param ver Version string to return.
 * @param len Maximum length of buffer.
 * @return Zero
 */
int model_version(char *ver, int len) {
	return coachella_version(ver, len);
}

int (*get_model_init())(const char *, const char *) {
        return &coachella_init;
}
int (*get_model_query())(coachella_point_t *, coachella_properties_t *, int) {
         return &coachella_query;
}
int (*get_model_finalize())() {
         return &coachella_finalize;
}
int (*get_model_version())(char *, int) {
         return &coachella_version;
}



#endif
